### PR TITLE
add VFGN utils and model code

### PR DESCRIPTION
**Pull request by dearleiii**
HP Virtual Foundry Graphnet - add VFGN utils code and model code

### CODE REVIEW

File: modulus/utils/vfgn/reading_utils.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+      tf.expand_dims(context['senders'], axis=0),
+      [input_trajectory_length, 1])
+
+  model_input_features['receivers'] = tf.tile(
+      tf.expand_dims(context['receivers'], axis=0),
+      [input_trajectory_length, 1])
+
+  # todo: change the hard-coded trajectory length to be the entire global context (/ sintering profile) sequence length
+  # sequence length here is the default sintering 2-stage total length
+  trajectory_length = 14+24
```
Comment (mnabian) [inline modulus/utils/vfgn/reading_utils.py:159]: Would be nice to get rid of the hard-coded sequence length here

File: modulus/models/vfgn/graph_network.py [hunk 2]
```diff
@@ -0,0 +1,562 @@
...
+        x_new = x + x_res
+        edge_attr_new = edge_attr + edge_attr_res
+
+        return x_new, edge_attr_new, receivers, senders
+
+
+class DecoderNet(torch.nn.Module):
+    def __init__(self, mlp_hidden_size, mlp_num_hidden_layers, output_size):
+        super(DecoderNet, self).__init__()
+        # self.node_mlp = MLPNet(mlp_hidden_size, mlp_num_hidden_layers, output_size)
```
Comment (mnabian) [inline modulus/models/vfgn/graph_network.py:177]: Please remove the commented out code.

File: modulus/models/vfgn/graph_network.py [hunk 3]
```diff
@@ -0,0 +1,562 @@
...
+            i += 1
+            if i % n_inter == 0:
+                j += 1
+
+        x = self._decoder_network(pre_x.to(origin_device))
+
+        return x
+
+
+class LearnedSimulator(torch.nn.Module):
```
Comment (mnabian) [inline modulus/models/vfgn/graph_network.py:245]: I suggest we move the data loading/processing modules to the `datapipes` folder to better organize the code.

File: modulus/models/vfgn/graph_network.py
Comment (mnabian) [inline modulus/models/vfgn/graph_network.py:1]: In general, there's a lot of overlap between this code and the MeshGraphNet code here: https://github.com/NVIDIA/modulus/blob/main/modulus/models/meshgraphnet/meshgraphnet.py. It is highly recommended to use the existing modules as much as possible and minimize code duplication.

**Review by mnabian — CHANGES_REQUESTED** [review]

File: modulus/models/vfgn/graph_network.py [hunk 4]
```diff
@@ -0,0 +1,562 @@
...
+import torch.nn.functional as F
+from torch_scatter import scatter
+from torch.utils.checkpoint import checkpoint
+
+
+STD_EPSILON = 1e-8
+
+
+class MLPNet(torch.nn.Module):
+    def __init__(self, mlp_hidden_size, mlp_num_hidden_layers, output_size, layer_norm=True):
```
[the remark this answers is not in the document]
Comment (dearleiii) [reply modulus/models/vfgn/graph_network.py:28]: discussed and agreed to keep the current implementation

**Pull request closed without merging**